x/ibc/testing/chain.go

```diff
@@ -0,0 +1,298 @@
+	// generate validator private/public key
+	privVal := tmtypes.NewMockPV()
+	pubKey, err := privVal.GetPubKey()
+	require.NoError(t, err)
+
+	// create validator set with single validator
+	validator := tmtypes.NewValidator(pubKey, 1)
+	valSet := tmtypes.NewValidatorSet([]*tmtypes.Validator{validator})
+	signers := []tmtypes.PrivValidator{privVal}
+
+	startTime := time.Now()
+
+	app := simapp.Setup(false)
+	ctx := app.BaseApp.NewContext(false,
+		abci.Header{
+			Height: 1,
+			Time:   startTime,
+		},
+	)
+
+	// generate and set senderAccount
+	senderPrivKey := secp256k1.GenPrivKey()
+	simapp.AddTestAddrsFromPubKeys(app, ctx, []crypto.PubKey{senderPrivKey.PubKey()}, sdk.NewInt(10000000000))
+	acc := app.AccountKeeper.GetAccount(ctx, sdk.AccAddress(senderPrivKey.PubKey().Address()))
+
+	// commit init chain changes so create client can be called by a counterparty chain
+	app.Commit()
+	// create current header and call begin block
+	header := abci.Header{
+		Height: 2,
+		Time:   time.Now(),
+	}
+	app.BeginBlock(abci.RequestBeginBlock{Header: header})
+
+	lastHeader := ibctmtypes.CreateTestHeader(clientID, 1, startTime, valSet, signers)
+
+	// create an account to send transactions from
+	return &TestChain{
+		t:             t,
+		ClientID:      clientID,
+		App:           app,
+		LastHeader:    lastHeader,
+		CurrentHeader: header,
+		Querier:       keeper.NewQuerier(*app.IBCKeeper),
+		Vals:          valSet,
+		Signers:       signers,
+		senderPrivKey: senderPrivKey,
+		senderAccount: acc,
+	}
+}
+
+// GetContext returns the current context for the application.
+func (chain *TestChain) GetContext() sdk.Context {
+	return chain.App.BaseApp.NewContext(false, chain.CurrentHeader)
+}
+
+// CommitNBlocks commits n blocks to state and updates the block height by 1 for each commit.
+func (chain *TestChain) CommitNBlocks(n int) {
+	for i := 0; i < n; i++ {
+		chain.App.Commit()
+		nextBlock(chain)
+	}
+
+}
+
+// CommitBlockWithNewTimestamp commits the current block and starts the next block with the
+// provided timestamp.
+func (chain *TestChain) CommitBlockWithNewTimestamp(timestamp int64) {
+	chain.App.Commit()
+	nextBlock(chain)
+	chain.App.BeginBlock(abci.RequestBeginBlock{
+		Header: abci.Header{
+			Height: chain.CurrentHeader.Height,
+			Time:   time.Unix(timestamp, 0),
+		}},
+	)
+}
+
+// nextBlock sets the last header to the current header and increments the current header to be
+// at the next block height and time.
+//
+// CONTRACT: this function must only be called after app.Commit() occurs
+func nextBlock(chain *TestChain) {
+	// set the last header to the current header
+	chain.LastHeader = ibctmtypes.CreateTestHeader(
+		chain.CurrentHeader.ChainID,
+		chain.CurrentHeader.Height,
+		chain.CurrentHeader.Time,
+		chain.Vals, chain.Signers,
+	)
+
+	// increment the current header
+	chain.CurrentHeader = abci.Header{
+		Height: chain.CurrentHeader.Height + 1,
+		Time:   time.Now(),
```

Comment:
thoughts on using `Now()` in testing? I'm not aware of how common non-deterministic bugs arise from this usage. 

I'm thinking the more robust approach is to add chain func to take in a counterparty chain and update the chains time and the counterparty's time by 5 seconds (or something like that). This limits this testing pkg to be used for 2 chains at a time only. 

This may be easier to implement if I allow references from this TestChain to the testing suite holding all the test chains

Reply:
Yea I like just taking the previous time and updating by like 5 seconds